openai_sdk_agent_app_sample_code/cookbook/agents/function_calling_agent.py

```diff
@@ -116,62 +116,72 @@ def predict(
-            model_response,
-            messages_log_with_tool_calls,
-        ) = self.recursively_call_and_run_tools(messages=messages)
-
-        # If your front end keeps of converastion history and automatically appends the bot's response to the messages history, remove this line.
-        messages_log_with_tool_calls.append(
-            model_response.choices[0].message.to_dict()
-        )  # OpenAI client
+        messages_log_with_tool_calls = self.recursively_call_and_run_tools(
+            messages=messages
+        )
 
         # remove the system prompt - this should not be exposed to the Agent caller
         messages_log_with_tool_calls = messages_log_with_tool_calls[1:]
 
         return {
-            "content": model_response.choices[0].message.content,
+            # "content": model_response.choices[0].message.content,
+            "content": messages_log_with_tool_calls[-1]["content"],
             # messages should be returned back to the Review App (or any other front end app) and stored there so it can be passed back to this stateless agent with the next turns of converastion.
             "messages": messages_log_with_tool_calls,
         }
 
-    @mlflow.trace(span_type="AGENT")
+    @mlflow.trace(span_type="CHAIN")
     def recursively_call_and_run_tools(self, max_iter=10, **kwargs):
         messages = kwargs["messages"]
         del kwargs["messages"]
         i = 0
         while i < max_iter:
-            response = self.chat_completion(messages=messages, tools=True)
-            assistant_message = response.choices[0].message  # openai client
-            tool_calls = assistant_message.tool_calls  # openai
-            if tool_calls is None:
-                # the tool execution finished, and we have a generation
-                return (response, messages)
-            tool_messages = []
-            for tool_call in tool_calls:  # TODO: should run in parallel
-                function = tool_call.function  # openai
-                args = json.loads(function.arguments)  # openai
-                result = execute_function(self.tool_functions[function.name], args)
-                tool_message = {
-                    "role": "tool",
-                    "tool_call_id": tool_call.id,
-                    "content": result,
-                }  # openai
-
-                tool_messages.append(tool_message)
-            assistant_message_dict = assistant_message.dict().copy()  # openai
-            del assistant_message_dict["content"]
-            del assistant_message_dict["function_call"]  # openai only
-            if "audio" in assistant_message_dict:
-                del assistant_message_dict["audio"]  # llama70b hack
-            messages = (
-                messages
-                + [
-                    assistant_message_dict,
-                ]
-                + tool_messages
-            )
-            i += 1
+            with mlflow.start_span(name=f"iteration_{i}", span_type="CHAIN") as span:
```

Comment:
Same question here RE "CHAIN" vs "AGENT"

Reply:
I mirrored what LangGraph traces looked like here - they use a CHAIN for this type of logic.  My thinking: this span represents logic that is actually a deterministic chain of steps.

Reply:
Got it, thanks! Do you have a pointer to a LangGraph "CHAIN" tracing example? i'm wondering if LangGraph uses that span name mostly writing a fixed chain, since the logic in this function calling agent thing looks similar to the Python equivalent of a LangGraph agent (e.g. if you translated something like LangGraph `create_react_agent` into Python code)